docs/scap.1
.\" Man page generated from reStructuredText.
.
.TH "SCAP" "1" "Feb 27, 2017" "3.0" "scap"
.SH NAME
scap \- Wikimedia deployment tool - scap: scatter crap around production
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.sp
These commands, for the most part, use rsync to synchronize the state of the
mediawiki config and wmf/* code branches from the deploy server to the
production web servers. There are also commands to perform other related
house\-keeping tasks as described in the individual command descriptions.
.SH SCAP SYNC
.sp
\fBscap sync\fP is the driver script for syncing the MediaWiki versions and
configuration files currently staged on the deploy server to the rest of the
cluster.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap sync [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                 [\-D <name>:<value>] [\-V] [\-v] [\-\-environment ENVIRONMENT]
                 [\-r] [\-\-force]
                 [message [message ...]]

Deploy MediaWiki to the cluster.

positional arguments:
  message               Log message for SAL

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-r, \-\-restart         Restart HHVM process on target hosts.
  \-\-force               Skip canary checks

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands,         
  not all commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap

#. Validate php syntax of wmf\-config and multiversion
#. Sync deploy directory on localhost with staging area
#. Create/update git repo in staging area
#. Compile wikiversions.json to php in deploy directory
#. Update l10n files in staging area
#. Compute git version information
#. Commit all changes to local git repo in deploy directory
#. Ask scap masters to sync with current master
#. Ask scap proxies to sync with master server
#. Ask apaches to sync with fastest rsync server
#. Ask apaches to rebuild l10n CDB files
#. Update wikiversions.php on localhost
#. Ask apaches to sync wikiversions.php
#. Restart HHVM across the cluster
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.Scap()\fP
.IP \(bu 2
\fBscap.tasks.check_php_syntax()\fP
.IP \(bu 2
\fBscap.tasks.compile_wikiversions()\fP
.IP \(bu 2
\fBscap.tasks.sync_common()\fP
.IP \(bu 2
\fBscap.tasks.sync_wikiversions()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH SCAP PULL
.sp
\fBscap pull\fP uses rsync to fetch MediaWiki code and configuration to the
local host. It is typically called automatically on hosts during the execution of \fI\%scap sync\fP\&.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap pull [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                 [\-D <name>:<value>] [\-V] [\-v] [\-\-environment ENVIRONMENT]
                 [\-\-no\-touch] [\-\-no\-update\-l10n] [\-i INCLUDE]
                 ...

Sync local MediaWiki deployment directory with deploy server state.

positional arguments:
  servers               Rsync server(s) to copy from

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-no\-touch            Do not touch InitialiseSettings.php after sync.
  \-\-no\-update\-l10n      Do not update l10n cache files.
  \-i INCLUDE, \-\-include INCLUDE
                        Rsync include pattern to limit transfer to. End
                        directories with a trailing \(ga/***\(ga. Can be used
                        multiple times.

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands, not all
  commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.SyncCommon()\fP
.IP \(bu 2
\fBscap.tasks.sync_common()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH SCAP SYNC-FILE
.sp
\fBscap sync\-file\fP synchronizes a file or directory from the staging
directory to the cluster.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap sync\-file [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                      [\-D <name>:<value>] [\-V] [\-v]
                      [\-\-environment ENVIRONMENT] [\-\-force]
                      [\-\-beta\-only\-change]
                      file [message [message ...]]

Sync a specific file/directory to the cluster.

positional arguments:
  file                  File/directory to sync
  message               Log message for SAL

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-force               Skip canary checks
  \-\-beta\-only\-change    Sync a config file that only affects beta cluster

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands, not all
  commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.SyncFile()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH SCAP SYNC-WIKIVERSIONS
.sp
\fBscap sync\-wikiversions\fP compiles wikiversions.json into a CDB database and then
syncs both the JSON and CDB versions to the rest of the cluster.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap sync\-wikiversions [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                              [\-D <name>:<value>] [\-V] [\-v]
                              [\-\-environment ENVIRONMENT] [\-\-force]
                              [message [message ...]]

Rebuild and sync wikiversions.php to the cluster.

positional arguments:
  message               Log message for SAL

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-force               Skip canary checks

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands, not all
  commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.SyncWikiversions()\fP
.IP \(bu 2
\fBscap.tasks.compile_wikiversions()\fP
.IP \(bu 2
\fBscap.tasks.sync_wikiversions()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH SCAP WIKIVERSIONS-INUSE
.sp
\fBscap wikiversions\-inuse\fP examines wikiversions.json to find the current active
MediaWiki versions.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap wikiversions\-inuse [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                               [\-D <name>:<value>] [\-V] [\-v]
                               [\-\-environment ENVIRONMENT] [\-\-withdb]

Get a list of the active MediaWiki versions.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-withdb              Add \(ga=wikidb\(ga with some wiki using the version.

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands, not all
  commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.MWVersionsInUse()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH SCAP WIKIVERSIONS-COMPILE
.sp
\fBwikiversions\-compile\fP compiles wikiversions.json into wikiversions.php.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap wikiversions\-compile [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                                 [\-D <name>:<value>] [\-V] [\-v]
                                 [\-\-environment ENVIRONMENT]

Compile wikiversions.json to wikiversions.php.

optional arguments:
  \-h, \-\-help            show this help message and exit

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands, not all
  commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.CompileWikiversions()\fP
.IP \(bu 2
\fBscap.tasks.compile_wikiversions()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH SCAP CDB-REBUILD
.sp
\fBscap cdb\-rebuild\fP rebuilds localization cache CDB files from JSON files.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap cdb\-rebuild [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                        [\-D <name>:<value>] [\-V] [\-v]
                        [\-\-environment ENVIRONMENT] [\-\-version VERSION]
                        [\-\-no\-progress] [\-\-staging]

Rebuild localization cache CDB files from the JSON versions.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-\-version VERSION     MediaWiki version (eg 1.27.0\-wmf.7)
  \-\-no\-progress         Do not show progress indicator.
  \-\-staging             Rebuild cdb files in staging directory

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands, not all
  commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.RebuildCdbs()\fP
.IP \(bu 2
\fBscap.tasks.merge_cdb_updates()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH SCAP CDB-JSON-REFRESH
.sp
\fBrefreshCdbJsonFiles\fP Create JSON/MD5 files for all CDB files in a directory.
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
usage: scap cdb\-json\-refresh [\-h] [\-c CONF_FILE] [\-\-no\-shared\-authsock]
                             [\-D <name>:<value>] [\-V] [\-v]
                             [\-\-environment ENVIRONMENT] \-d DIRECTORY
                             [\-t THREADS]

Create JSON/MD5 files for all CDB files in a directory.

optional arguments:
  \-h, \-\-help            show this help message and exit
  \-d DIRECTORY, \-\-directory DIRECTORY
                        Directory containing cdb files
  \-t THREADS, \-\-threads THREADS
                        Number of threads to use to build json/md5 files

global arguments:
  Although these arguments can be passed to all scap (sub\-)commands,         
  not all commands are affected by every global argument.

  \-c CONF_FILE, \-\-conf CONF_FILE
                        Path to configuration file
  \-\-no\-shared\-authsock  Ignore any shared ssh\-auth configuration
  \-D <name>:<value>, \-\-define <name>:<value>
                        Set a configuration value
  \-V                    Print scap version.
  \-v, \-\-verbose         Verbose output
  \-\-environment ENVIRONMENT
                        environment in which to execute scap

This will put a JSON and MD5 file in /upstream for each CDB file.

This can be combined with rsync and the scap\-rebuild\-cdbs to
push out large CDB files with minimal traffic. CDB files change
drastically with small key/value changes, where as JSON files do not, and
thus they diff/rdiff much better.

When pushing updates with rsync, this should be run before running rsync.
The rsync command should exclude CDB files or at least use
\-ignore\-existing. After the rsync is done, scap\-rebuild\-cdbs can be
run on each server to apply the updates to the CDB files.
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
\fBscap.refreshCdbJsonFiles()\fP
.IP \(bu 2
\fBscap.tasks.refresh_cdb_json_files()\fP
.UNINDENT
.UNINDENT
.UNINDENT
.SH COPYRIGHT
2014-2017, Wikimedia Foundation & Contributors.
Released under the terms of the GNU General Public License, version 3 <https://www.gnu.org/licenses/gpl-3.0.txt>.
.\" Generated by docutils manpage writer.
.


changelog
scap (4.46.0)

  * Exclude deploy servers from target list if /srv/mediawiki is a symlink
  * kubernetes.py: Extract HELM_* env vars from /etc/profile.d/kube-env.sh if available
  * Remove an obsolete comment

 -- Ahmon Dancy <adancy@wikimedia.org>  Mon, 13 Mar 2023 10:00:40 -0700

scap (4.45.0)

  * Remove l10nupdate support

 -- Ahmon Dancy <adancy@wikimedia.org>  Fri, 10 Mar 2023 08:02:33 -0800

scap (4.44.0)

  * fix typo
  * install-world: add `exclude-hosts` regex argument
  * targets: add convenience `exclude_hosts` regex argument
  * cleanup: remove leftover PyInstaller code

 -- jnuche <jnuche@wikimedia.org>  Wed, 08 Mar 2023 18:30:14 +0100

scap (4.43.0)

  * Check for PHP notices during mwscript mergeMessageFileList.php
  * Make scap pull exit if /srv/mediawiki is a symlink

 -- Ahmon Dancy <adancy@wikimedia.org>  Tue, 28 Feb 2023 11:45:48 -0800

scap (4.42.0)

  * self-installer: overhaul installation mechanism

 -- jnuche <jnuche@wikimedia.org>  Tue, 28 Feb 2023 10:30:01 +0100

scap (4.41.0)

  * .gitlab-ci.yml: fix name of published images

 -- jnuche <jnuche@wikimedia.org>  Fri, 24 Feb 2023 10:34:10 +0100

scap (4.40.0)

  * .gitlab-ci.yml: fix matrix of `publish-dist` job

 -- jnuche <jnuche@wikimedia.org>  Thu, 23 Feb 2023 17:46:30 +0100

scap (4.39.0)

  * binary distributions: use Python wheels to distribute Scap

 -- jnuche <jnuche@wikimedia.org>  Thu, 23 Feb 2023 16:03:21 +0100

scap (4.38.0)

  * Discard stderr noise generated by helm/helmfile

 -- Ahmon Dancy <adancy@wikimedia.org>  Tue, 21 Feb 2023 10:19:19 -0800

scap (4.37.0)

  * Scap backport: Request dependencies for a specific change
  * Remove support for deprecated DOLOGMSGNOLOG env var

 -- Jeena Huneidi <jhuneidi@wikimedia.org>  Thu, 16 Feb 2023 15:50:59 -0800

scap (4.36.0)

  * Attempt to autorecover from interrupted helm deployments

 -- Ahmon Dancy <adancy@wikimedia.org>  Thu, 16 Feb 2023 10:40:08 -0800

scap (4.35.0)

  * Don't transmit "aborted" message to IRC if no prior announcement has been made
  * scap say: Fix longstanding handling of COLUMNS env var
  * Perform k8s deployment before proxies/apaches
  * Load section config when host has a trailing dot
  * Improve scap.config test coverage
  * tox: add a virtualenv for local testing

 -- Ahmon Dancy <adancy@wikimedia.org>  Wed, 15 Feb 2023 13:06:06 -0800

scap (4.34.0)

  * Replace "universal_newlines" with "text"
  * deploy-promote: Fix how prev_version is determined

 -- Ahmon Dancy <adancy@wikimedia.org>  Thu, 02 Feb 2023 08:46:11 -0800

scap (4.33.1)

  * Add binary image support

 -- Ahmon Dancy <adancy@wikimedia.org>  Thu, 19 Jan 2023 13:30:34 -0800

scap (4.33.0)

  * Debug log command used when self-restarting php-fpm
  * Avoid duplicate usernames in "synced to the testservers" message
  * requirements.txt: packaging==23
  * kubernetes.py: Use correct image for debug envs
  * Remove .pipeline/config.yaml

 -- Ahmon Dancy <adancy@wikimedia.org>  Wed, 18 Jan 2023 08:42:17 -0800

scap (4.32.0)

  * backport: Don't merge already merged patches

 -- Ahmon Dancy <adancy@wikimedia.org>  Wed, 04 Jan 2023 11:04:35 -0800

scap (4.31.1)

  * kubernetes.py: Ensure log messages aren't too large

 -- Ahmon Dancy <adancy@wikimedia.org>  Wed, 04 Jan 2023 09:35:16 -0800

scap (4.31.0)

  * Remove a now-resolved FIXME comment
  * Send image build/deploy transcripts to syslog
  * Add full_image_build config option

 -- Ahmon Dancy <adancy@wikimedia.org>  Wed, 04 Jan 2023 07:47:49 -0800

scap (4.30.3)

  * RELEASE.md: Change "Gerrit" to "Gitlab"
  * scap backport: Change-Id is case insensitive
  * use trusted tag instead of protected

 -- Jeena Huneidi <jhuneidi@wikimedia.org>  Fri, 16 Dec 2022 14:14:32 -0800

scap (4.30.2)

  * perform-release: do not push all local tags
  * kubernetes: restore correct rollback behavior for deployments

 -- jnuche <jnuche@wikimedia.org>  Fri, 09 Dec 2022 18:55:19 +0100

scap (4.30.1)

  * requirements: update dependencies versions to fix Gitlab pipeline
  * kubernetes: treat operation failures as soft errors

 -- jnuche <jnuche@wikimedia.org>  Fri, 09 Dec 2022 10:51:33 +0100

scap (4.30.0)

  * pull MV image onto K8s cluster before any actual deployments
  * remove command line options for K8s deployments
  * Add support for "gerriturl/r/:changenum" URLs to scap backport command

 -- Ahmon Dancy <adancy@wikimedia.org>  Thu, 01 Dec 2022 09:56:45 -0800

scap (4.29.3)

  * perform-release: display current version when prompting for new one
  * utils.find_nearest_host: correctly catch timeouts

 -- Brennen Bearnes <bbearnes@wikimedia.org>  Tue, 29 Nov 2022 15:12:50 -0700

scap (4.29.2)

  * k8s: allow to perform full K8s deployments with `deploy-promote`

 -- jnuche <jnuche@wikimedia.org>  Mon, 28 Nov 2022 17:53:48 +0100

scap (4.29.1)

  * kubernetes: Fix SAL suppression

 -- jnuche <jnuche@wikimedia.org>  Tue, 22 Nov 2022 13:31:22 +0100

scap (4.29.0)

  * kubernetes: allow pre-pulling the multiversion image on k8s nodes
  * blubber.yaml: Fix syntax line
  * kubernetes: Do not log to SAL
  * utils: Allow env for subprocess_check_run_quietly_if_ok
  * perform-release: remove support for Gerrit remote
  * Release 4.28.2-1

 -- jnuche <jnuche@wikimedia.org>  Mon, 21 Nov 2022 13:05:36 +0100

scap (4.28.2)

  * kubernetes: do not run "helmfile test"

 -- Ahmon Dancy <adancy@wikimedia.org>  Thu, 17 Nov 2022 08:20:15 -0800

scap (4.28.1)

  * kubernetes: only deploy to the selected release

 -- Ahmon Dancy <adancy@wikimedia.org>  Mon, 14 Nov 2022 09:08:22 -0800

scap (4.28.0)

  * k8s: add deployment stages for canaries and production
  * k8s: Pass mv_debug_image_name when building container images
  * perform-release: Add GitLab support
  * Update blubber buildkit image references
  * Publish scap docs when merging into the default branch
  * .dockerignore: Add Makefile
  * Replace some references to gerrit with gitlab
  * bin/install_local_version.sh: Update verify_source_dir for GitLab

subplot/scap.py
# Note: This is not a self-standing Python file. It gets embedded into a
# generated program by Subplot. That is why any functions that will be in the
# full program are accessed by retrieving them via the globals() dict.
#
# Part of the interface for step functions expected by Subplot is that any
# parts of steps captured by bindings (see scap.yaml) are given to the step
# functions as keyword arguments. The step functions "declare" what they expect
# with 'foo=None' in the function definition.

import logging
import os
import random
import re


def nop(ctx, *args, **kwargs):
    pass


def mkdir(ctx, pathname=None):
    os.makedirs(pathname)


def file_from(ctx, filename=None, source=None):
    get_file = globals()["get_file"]

    dirname = os.path.dirname(filename) or "."
    logging.debug("file_from: filename=%s", filename)
    logging.debug("file_from: dirname=%s", dirname)
    if not os.path.exists(dirname):
        os.makedirs("./" + dirname)

    _write(filename, get_file(source) + b"\n")


def git_init(ctx, pathname=None):
    _runcmd(ctx, ["git", "init", pathname])
    exit_code_is(ctx, 0)

    # Write a random number to a file to make sure commit ids are
    # always different.
    _write(
        os.path.join(pathname, "dummy"), str(random.randint(0, 2 ** 62)).encode("UTF-8")
    )

    # Add the file so the repo isn't empty and has a master branch.
    _runcmd(ctx, ["git", "add", "."], cwd=pathname)
    _runcmd(ctx, ["git", "commit", "-mdummy", "dummy"], cwd=pathname)
    _save_git_tip(ctx, pathname, "master")


def git_commit_file(ctx, source=None, pathname=None):
    git_commit_file_from(ctx, filename=source, source=source, pathname=pathname)


def git_commit_file_from(ctx, filename=None, source=None, pathname=None):
    get_file = globals()["get_file"]
    _write(os.path.join(pathname, filename), get_file(source) + b"\n")

    _runcmd(ctx, ["git", "add", filename], cwd=pathname)
    exit_code_is(ctx, 0)

    _runcmd(ctx, ["git", "commit", "-qm", source], cwd=pathname)
    exit_code_is(ctx, 0)

    _save_git_tip(ctx, pathname, "master")


def git_am(ctx, dirname=None, filename=None):
    _runcmd(ctx, ["git", "am", filename], cwd=dirname)


def git_am_abort(ctx, dirname=None):
    _runcmd(ctx, ["git", "am", "--abort"], cwd=dirname)


def git_am_in_progress(ctx, pathname=None):
    _runcmd(ctx, ["git", "am", "--abort"], cwd=pathname)
    stdout = ctx["stdout"]
    assert ctx["exit"] != 0
    assert "git am" in stdout


def run_scap_subcommand_help(ctx, subcommand=None):
    _scap(ctx, [subcommand, "--help"])
    runcmd_exit_code_is = globals()["runcmd_exit_code_is"]
    runcmd_exit_code_is(ctx, 0)


def run_scap(ctx):
    _scap(ctx, [])


def run_scap_version(ctx):
    _scap(ctx, ["version"])


def run_scap_sync(ctx):
    _scap(ctx, ["sync"])


def run_scap_apply_patches(ctx, code=None, patches=None, train=None):
    _scap(
        ctx,
        [
            "apply-patches",
            f"-Dstage_dir:{code}",
            f"-Dpatch_path:{patches}",
            "--train",
            train,
        ],
    )


def run_scap_test_patches(ctx, code=None, patches=None, train=None):
    _scap(
        ctx,
        [
            "test-patches",
            f"-Dstage_dir:{code}",
            f"-Dpatch_path:{patches}",
            "--train",
            train,
        ],
    )


def run_scap_list_patches(ctx, patches=None, train=None):
    _scap(ctx, ["list-patches", f"-Dpatch_path:{patches}", "--train", train])


def git_working_tree_is_clean(ctx, pathname=None):
    assert_eq = globals()["assert_eq"]
    exit, lines = git_working_tree_status(ctx, pathname=pathname)
    assert_eq(exit, 0)
    for line in lines:
        assert "You are in the middle of an am session." not in line


def git_working_tree_is_dirty(ctx, pathname=None):
    assert_ne = globals()["assert_ne"]
    exit, lines = git_working_tree_status(ctx, pathname=pathname)
    assert_ne(lines, [])


def git_working_tree_status(ctx, pathname=None):
    runcmd_get_stdout = globals()["runcmd_get_stdout"]
    runcmd_get_exit_code = globals()["runcmd_get_exit_code"]

    _runcmd(ctx, ["git", "status", "--ignore-submodules"], cwd=pathname)
    exit = runcmd_get_exit_code(ctx)
    stdout = runcmd_get_stdout(ctx)
    lines = stdout.splitlines()
    lines = [x for x in lines if x != "?? extensions/"]
    logging.debug("tree status: %r", lines)
    return exit, lines


def repo_has_changed(ctx, pathname=None):
    assert_ne = globals()["assert_ne"]
    assert_ne(_get_git_tip(pathname, "master"), _remembered_tip(ctx, pathname))


def repo_has_not_changed(ctx, pathname=None):
    assert_eq = globals()["assert_eq"]
    assert_eq(_get_git_tip(pathname, "master"), _remembered_tip(ctx, pathname))


def repo_change_matches(ctx, pathname=None, filename=None):
    get_file = globals()["get_file"]
    runcmd_get_stdout = globals()["runcmd_get_stdout"]

    # Check that the changes since the remembered tip match a patch.
    _runcmd(ctx, ["git", "diff", _remembered_tip(ctx, pathname), "HEAD"], cwd=pathname)
    exit_code_is(ctx, 0)

    diff = runcmd_get_stdout(ctx)
    _write("diff", diff.encode("UTF-8"))
    _write(filename, get_file(filename))
    _runcmd(ctx, ["interdiff", "diff", filename], input=diff)
    exit_code_is(ctx, 0)
    stdout_is_empty(ctx)


def run_scap_sync_world(ctx):
    scap = _binary("bin", "scap")
    _runcmd(ctx, [scap, "sync-world"])


def exit_code_is(ctx, exit_code=None):
    runcmd_exit_code_is = globals()["runcmd_exit_code_is"]
    runcmd_exit_code_is(ctx, exit_code)


def stdout_is_empty(ctx):
    assert_eq = globals()["assert_eq"]
    runcmd_get_stdout = globals()["runcmd_get_stdout"]

    stdout = runcmd_get_stdout(ctx)
    assert_eq(stdout, "")


def stdout_matches(ctx, pattern=None):
    assert_ne = globals()["assert_ne"]
    runcmd_get_stdout = globals()["runcmd_get_stdout"]

    stdout = runcmd_get_stdout(ctx)
    m = re.search(pattern, stdout, flags=re.M)
    if not m:
        logging.debug("stdout: %r" % repr(stdout))
        logging.debug("pattern: %r" % repr(pattern))
    assert_ne(m, None)


def stdout_matches_and_later(ctx, pattern1=None, pattern2=None):
    assert_ne = globals()["assert_ne"]
    runcmd_get_stdout = globals()["runcmd_get_stdout"]

    output = runcmd_get_stdout(ctx)
    m1 = re.search(pattern1, output)
    assert_ne(m1, None)

    m2 = re.search(pattern2, output[m1.end() :])
    assert_ne(m2, None)


def stdout_does_not_match(ctx, pattern=None):
    assert_eq = globals()["assert_eq"]
    runcmd_get_stdout = globals()["runcmd_get_stdout"]

    stdout = runcmd_get_stdout(ctx)
    m = re.search(pattern, stdout)
    assert_eq(m, None)


def stderr_matches(ctx, pattern=None):
    assert_ne = globals()["assert_ne"]
    runcmd_get_stderr = globals()["runcmd_get_stderr"]

    stderr = runcmd_get_stderr(ctx)
    m = re.search(pattern, stderr)
    assert_ne(m, None)


def _binary(*basenames):
    srcdir = globals()["srcdir"]
    return os.path.join(srcdir, *basenames)


def _scap(ctx, args):
    srcdir = globals()["srcdir"]

    scap = _binary("bin", "scap")
    env = dict(os.environ)
    env["PYTHONPATH"] = srcdir
    if "SSH_AUTH_SOCK" in env:
        del env["SSH_AUTH_SOCK"]
    _runcmd(ctx, [scap] + args, env=env)


def _runcmd(ctx, argv, cwd=None, env=None, input=""):
    if env is None:
        env = {}
    runcmd_run = globals()["runcmd_run"]
    logging.debug("_runcmd: cwd={}".format(cwd))
    runcmd_run(ctx, argv, cwd=cwd, env=env)


def _write(filename, content):
    dirname = os.path.dirname(filename)
    if dirname and not os.path.exists(dirname):
        os.makedirs(dirname)
    with open(filename, "wb") as f:
        f.write(content)


def _save_git_tip(ctx, pathname, ref):
    tip = _get_git_tip(pathname, ref)
    tips = ctx.get("tips", {})
    tips[pathname] = tip
    ctx["tips"] = tips


def _get_git_tip(pathname, ref):
    runcmd_get_stdout = globals()["runcmd_get_stdout"]

    Context = globals()["Context"]

    dummy = Context()
    _runcmd(dummy, ["git", "rev-parse", ref], cwd=pathname)
    exit_code_is(dummy, 0)
    stdout = runcmd_get_stdout(dummy)
    return stdout.strip()


def _remembered_tip(ctx, pathname):
    ret = ctx["tips"][pathname]
    return ret


def remove_file(ctx, filename=None):
    os.remove(filename)
